### DISCUSSION TITLE
retry every n minutes if internet connection is not available

### DISCUSSION THREAD

**Discussion by imed-ghomari** [Taitava/obsidian-shellcommands#168] [Q&A] [answered]
Like my previous question, I don't know if this can already be done or is yet to be implemented. 
But, for shell commands that require internet connection, is there a way to retry every 5 minutes until connection is restored ?

**Comment by Taitava** [comment, marked as answer]
Hi, that's a good question!

The plugin itself does not have any mechanism for trying execution again every few minutes if the execution fails. What you could do however, is to write a shell command that contains a loop that tries to call a command multiple times until it succeeds. The loop would have a `sleep` command too, that would introduce a five minutes pause between command calls, and also a counter that stops the loop in case the network connection does not recover after e.g. 10 tries (= 10 * 5 minutes = trying to call the command for 50 minutes).

May I ask you what shell you are using? [Here is a guide on how to find out what shell you are using if you don't know it](https://publish.obsidian.md/shellcommands/Operating+systems+and+shells/Shells#How+to+know+which+shell+is+used). In case you happen to be using Bash, you could have the following kind of shell command set:
```bash
for counter in {1..10}; do echo "Try #"$counter"..."; wget "https://github.com" -O "outputfile.html"; if [$? -eq 0]; then break; fi; echo "Failed, will sleep for 5 minutes before retrying."; sleep 5m; done
```
- I'm using `wget "https://github.com" -O "outputfile.html"` command to download something from the internet, but you can change this command to whatever command you want to use that does require an internet connection.
- If `counter` raises over 10, the `for` loop will automatically exit.
- `if [$? -eq 0]; then break; fi;` checks if the command succeeded or not. If the last command's exit code is 0, then it's considered as succeeded (= no error) and the loop will exit with the `break` command. If the exit code is something else than 0, then it's an error number meaning that some failure has happened.
- `sleep 5m` introduces a pause of 5 minutes. If you want to shorten it to e.g. 30 seconds, use `sleep 30s`.
- If your shell is something else than Bash, just let me know and I'll try to adapt the command to work on that.

## No output during the process

A not so optimal thing in this is that when the loop can take a long time to run, you cannot easily monitor how the process is going. If you `echo` something during the process (like is done in the example), you will only receive the echoed output to Obsidian **after** the whole loop ends. Like mentioned before, if the amount of tries is ten, and pause duration is five minutes, then it might take up to 50 minutes until you'll receive the echoed texts to Obsidian. This is something that #64 will try to solve sometime, but probably not anytime soon.

One workaround is of course to use `echo "Try #"$counter".." >> SomeNoteFile.md` to echo text into a file during the process. If you have that file open in Obsidian, then you'll see realtime text output, but using a file for this can be a bit cumbersome.

## Make the shell command easier to read

The shell command is quite long as a single line version. ~~The SC plugin does [not yet support multiline shell commands](#153), but if you want to make it multiline already now, you can save the shell command in a file named e.g. `my-internet-command.sh` and use the following content in it:~~ Edit 2022-07-02: Since SC version `0.13.0`, the plugin now supports pressing enter in shell command fields, so you can put the following straight into the shell command field, no need for a separate script file anymore:
```bash
for counter in {1..10};
do
  echo "Try #"$counter"..."; wget "https://github.com" -O "outputfile.html";
  if [$? -eq 0];
  then
    break;
  fi;
  echo "Failed, will sleep for 5 minutes before retrying.";
  sleep 5m;
done
```

~~If you store the file in your vault's root directory, you can use the following shell command to call the script file: `/bin/bash my-internet-command.sh`. If it's in a subfolder, then e.g.: `/bin/bash my-scripts/my-internet-command.sh`.~~

**Reply by imed-ghomari** [reply]
Yes, I'm using bash, appreciate you taking the time to write a solution.
> Make the shell command easier to read

I didn't know that, thanks for mentioning this.

**Reply by Taitava** [reply]
## Multiline shell commands are now supported
I just edited my answer slightly to indicate that currently, since SC version `0.13.0`, multiline shell commands are supported, so there is no more need for a separate script file. You can just press enter in any shell command field to make it multiline, or copy and paste the multiline example straight to the shell command field.
